Sage-Goddess/odoo
# 10 ss app extension

### Which issue does this close?

Closes #10 

### Write a short description of code change.

The engine behind Order Syncing (Cancels, Ships, QTY Change, Split Orders and more) and Posting Inventory Data (Location and QTY On hand) to ShipStation.  View [documentation](https://github.com/Sage-Goddess/odoo/wiki/Shipstation-Integration) 👀 

### For every PR please do the following:

* [x] Did you follow [Odoo coding standards](https://www.odoo.com/documentation/14.0/developer/misc/other/guidelines.html)?
* [x] Was all debug code removed?

### List steps and scenarios to test these changes.
  
Remember any database state or configuration steps needed.

Steps to test:
1. Update App list in odoo dash, and install Base Shipstation Module then SG Shipstation Module
2. Valdate configuration settings in ir_config_parameter - for staging ensure 'ss_inv_disable_call_to_ss' is 1 to prevent modifying shipstation production
3. Go to Sales > Configuration > Shipstation STores and import stores
4. Remove unused stores
5. Set up 2 stores 1 for awating_shipment and another for `cancelled`, set Warehouse
6. Open awaiting_shipment store and begin manually importing orders from shipstation till satisfied
7. Now, for both stores, enable Add To automatic shipstation
8. Turn on Order Sync Cron, Turn on Shipment Cron, Turn on Inventory Data Cron
9. Monitor logs, shipstation_operation_details and listed database tables for flow:
    - stock_picking (shipments/do), sale_order, sg_ships_inventory (inventory data)

### Convenience queries:

#### Inventory Cron Query
```sql
-- Inventory Cron Pickup
SELECT
	id, ss_store_id, order_id, order_number, order_date, order_status, status, failed_attempts, create_date, last_touched_at, write_date
FROM
	"sg_shipstation_inventory"
WHERE ((((("sg_shipstation_inventory"."status" = 'pending')
				OR("sg_shipstation_inventory"."status" in('fail_network', 'fail_bad_request', 'fail_empty_response', 'fail_sku', 'fail_general')))
			AND(("sg_shipstation_inventory"."order_status" NOT in('shipped', 'cancelled'))
			OR "sg_shipstation_inventory"."order_status" IS NULL))
		AND("sg_shipstation_inventory"."failed_attempts" <= 3)) -- inventory_failed_attempts
	AND("sg_shipstation_inventory"."last_touched_at" IS NULL
		OR("sg_shipstation_inventory"."last_touched_at" <= timezone('utc', now()) - INTERVAL '1 hour'))) -- ss_inv_last_touched_interval
ORDER BY
	"sg_shipstation_inventory"."write_date" DESC -- DESC Newest to oldest nulls to top, ASC Oldest to Newest nulls to bottom
LIMIT 30 -- inventory_amount_to_process
```

#### Shipments: Active Hours 6am - 7pm - Morning, Afternoon, Evening, Night
```sql
-- Shipped: ACTIVE HOURS: 6am - 7pm - Morning, Afternoon, Evening, Night
SELECT
	"stock_picking".id
	, create_date
	, last_touched_at
	, "state"
	, ss_overall_check_count as "Overall Count"
	, ss_not_shipped_check_count as "Not Ship"
	, ss_has_shipment_check_count as "Has Ship"
	, shipstation_order_id as "SS Order ID"
	, shipstation_sale_order_number as "SS Order #"
	, timezone('utc', now()) - INTERVAL '0 days' as "Now UTC"
	, timezone('utc', now()) - INTERVAL '7 days' as "-7 Days (From Date)"
	, timezone('utc', now()) - INTERVAL '1 hour' as "-1 Hour (Last Touched)"
FROM
	"stock_picking"
WHERE (((((("stock_picking"."state" = 'assigned')
					AND("stock_picking"."is_order_imported_in_shipstation" IS NULL
						OR "stock_picking"."is_order_imported_in_shipstation" = FALSE))
				AND "stock_picking"."shipstation_order_id" IS NOT NULL)
			AND("stock_picking"."picking_type_id" in(
					SELECT
						"stock_picking_type".id FROM "stock_picking_type"
					WHERE ("stock_picking_type"."code" = 'outgoing')
				ORDER BY
					"stock_picking_type"."id")))
		AND("stock_picking"."create_date" >= timezone('utc', now()) - INTERVAL '7 days')) -- NOW -7 DAYS = timezone('utc', now()) - INTERVAL '7 days'  shipped_from_date_in_hours
	AND("stock_picking"."last_touched_at" IS NULL
		OR("stock_picking"."last_touched_at" <= timezone('utc', now()) - INTERVAL '1 hour'))) -- ss_last_touched_interval
ORDER BY
	"stock_picking"."last_touched_at" ASC -- DESC Newest to oldest nulls to top, ASC Oldest to Newest nulls to bottom
LIMIT 20 -- shipped_amount_to_process
```

#### Shipped: After Hours - 7pm to 5am
```sql
-- AFTER HOURS: 7pm to 5am
SELECT
	"stock_picking".id
	, create_date
	, last_touched_at
	, shipstation_order_id as "SS Order ID"
	, shipstation_sale_order_number as "SS Order #"
	, timezone('utc', now()) - INTERVAL '0 days' as "Now UTC"
	, timezone('utc', now()) - INTERVAL '120 days' as "-120 Days (Cutoff)"
	, timezone('utc', now()) - INTERVAL '7 days' as "-7 Days (From Date)"
	, timezone('utc', now()) - INTERVAL '1 hour' as "-1 Hour (Last Touched)"
FROM
	"stock_picking"
WHERE ((((((("stock_picking"."state" = 'assigned')
						AND("stock_picking"."is_order_imported_in_shipstation" IS NULL
							OR "stock_picking"."is_order_imported_in_shipstation" = FALSE))
					AND "stock_picking"."shipstation_order_id" IS NOT NULL)
				AND("stock_picking"."picking_type_id" in(
						SELECT
							"stock_picking_type".id FROM "stock_picking_type"
						WHERE ("stock_picking_type"."code" = 'outgoing')
					ORDER BY
						"stock_picking_type"."id")))
			AND("stock_picking"."create_date" <= timezone('utc', now()) - INTERVAL '7 days')) -- NOW MINUS 7 = timezone('utc', now()) - INTERVAL '7 days'
		AND("stock_picking"."create_date" >= timezone('utc', now()) - INTERVAL '120 days')) -- NOW MINUS 120 DAYS/4 MONTHS = timezone('utc', now()) - INTERVAL '120 days'
	AND("stock_picking"."last_touched_at" IS NULL
		OR("stock_picking"."last_touched_at" <= timezone('utc', now()) - INTERVAL '1 hours'))) -- NOW MINUS 1 HOUR = timezone('utc', now()) - INTERVAL '1 hour'
ORDER BY
	"stock_picking"."last_touched_at" DESC -- Oldest First
LIMIT 20
```
